tests/unit/test_lib_per_app_requires.py

```diff
@@ -89,3 +90,69 @@ def test_validator(requirer: IngressPerAppRequirer, harness, auto_data, ok):
     else:
         host, port = auto_data
         requirer.provide_ingress_requirements(host=host, port=port)
+
+
+class TestIPAEventsEmission(unittest.TestCase):
+    class _RequirerCharm(CharmBase):
+        META = dedent(
+            """\
+            name: ipa-requirer
+            requires:
+              ingress:
+                interface: ingress
+                limit: 1
+            """
+        )
+
+        ready_event_count: int = 0
+        revoked_event_count: int = 0
+
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.ipa = IngressPerAppRequirer(self, relation_name="ingress", port=80)
+            self.framework.observe(self.ipa.on.ready, self._on_ready)
+            self.framework.observe(self.ipa.on.revoked, self._on_revoked)
+
+        def _on_ready(self, _event):
+            self.ready_event_count += 1
+
+        def _on_revoked(self, _event):
+            self.revoked_event_count += 1
+
+    def setUp(self):
+        self.harness = Harness(self._RequirerCharm, meta=self._RequirerCharm.META)
+        self.addCleanup(self.harness.cleanup)
+
+        self.harness.set_model_name(self.__class__.__name__)
+        self.harness.begin_with_initial_hooks()
+
+    def test_ipa_events(self):
+        # WHEN an ingress relation is formed
+        before = self.harness.charm.ready_event_count
+        self.rel_id = self.harness.add_relation("ingress", "traefik-app")
+        self.harness.add_relation_unit(self.rel_id, "traefik-app/0")
+
+        # AND an ingress is in effect
+        data = {"url": "http://a.b/c"}
+        self.harness.update_relation_data(
+            self.rel_id, "traefik-app", {"ingress": yaml.safe_dump(data)}
+        )
+        self.assertEqual(self.harness.charm.ipa.url, "http://a.b/c")
+
+        # THEN the ready event is emitted
+        after = self.harness.charm.ready_event_count
```

Comment:
this seems like a good place to use https://discourse.charmhub.io/t/harness-recipe-capture-events/6581

Reply:
Done. This is magic!